src/mlpack/methods/lsh/lshmodel_impl.hpp

```diff
@@ -0,0 +1,607 @@
+/**
+ * @file lshmodel_impl.hpp
+ * @author Yannis Mentekidis
+ *
+ * Implementation of the LSHModel functions.
+ */
+#ifndef MLPACK_METHODS_NEIGHBOR_SEARCH_LSH_MODEL_IMPL_HPP
+#define MLPACK_METHODS_NEIGHBOR_SEARCH_LSH_MODEL_IMPL_HPP
+
+#include "lshmodel.hpp"
+#include <boost/math/distributions/normal.hpp> // pdf and cdf needed
+
+
+//TODO: remove
+using std::cout;
+using std::flush;
+using std::endl;
+
+namespace mlpack {
+namespace neighbor {
+
+// Constructor sets variables and trains the object.
+template <typename SortPolicy, typename ObjectiveFunction>
+LSHModel<SortPolicy, ObjectiveFunction>::
+LSHModel(const arma::mat &referenceSet,
+         const double sampleSize,
+         const size_t k)
+{
+  // We don't own the set - we just point to it.
+  ownsSet = false;
+  this->referenceSet = &referenceSet;
+
+  Train(referenceSet, sampleSize, k);
+}
+
+// Destructor must de-allocate any referenceSet and LSHSearch objects we own.
+template <typename SortPolicy, typename ObjectiveFunction>
+LSHModel<SortPolicy, ObjectiveFunction>::~LSHModel()
+{
+  if (ownsSet)
+    delete referenceSet;
+};
+
+// Trains the object.
+template <typename SortPolicy, typename ObjectiveFunction>
+void LSHModel<SortPolicy, ObjectiveFunction>::Train(
+    const arma::mat &referenceSet,
+    const double sampleRate,
+    const size_t k)
+{
+  // Sanity check - sample rate must be in (0, 1].
+  if (sampleRate > 1 || sampleRate <= 0)
+    Log::Fatal << "Sampling rate must be floating point number in (0, 1]"
+        << std::endl;
+
+  // Update the object's max K value information.
+  maxKValue = k;
+
+  // Save pointer to training set.
+  this->referenceSet = &referenceSet;
+
+  // Step 1. Select a random sample of the dataset. We will work with only that
+  // sample.
+  arma::vec sampleHelper(referenceSet.n_cols, arma::fill::randu);
+
+  // Keep a sample of the dataset: We have uniformly random numbers in [0, 1],
+  // so we expect about N*sampleRate of them to be in [0, sampleRate).
+  arma::mat sampleSet = referenceSet.cols(
+        arma::find(sampleHelper < sampleRate));
+  // Shuffle to be impartial (in case dataset is sorted in some way).
+  sampleSet = arma::shuffle(sampleSet);
+  const size_t numSamples = sampleSet.n_cols; // Points in sampled set.
```

Comment:
I think it's without replacement: I generate uniform numbers in [0, 1] and then threshold at the sample rate, getting a vector of booleans. I keep only columns (so, points) that have "true" in the corresponding vector position:
In matlab/pseudocode it would be:
```MATLAB
sampleRate = 0.3;
referenceSet = [
1 3 5 7;
2 4 6 8;
]
sampleHelper = [0.1 0.3 0.7 0.05];
sampleHelper = sampleHelper > sampleRate;
% So here sampleHelper = [0 0 1 0]
sampleSet = referenceSet.cols(sampleHelper);
% and therefore sampleSet = [5; 6] - only column 3
```

Is there something I don't see here?

I didn't know about `ObtainDistinctSamples()`, I think that will make the code cleaner so I'll refactor it to use that instead.